File: packages/rune-safe-area/ios/RuneSafeArea.swift
//
//  RuneSafeArea.swift
//  RuneSafeArea
//
//  Public interface for RuneSafeArea module
//

import Foundation
import RuneKit

@objc(RuneSafeArea)
public class RuneSafeArea: NSObject {
  
  private static var moduleInstance: RuneSafeAreaModule?
  
  /// Initialize the safe area module with a RuneRuntime instance
  /// Should be called during app initialization before the JS bundle loads
  @objc public static func initialize(with runtime: RuneRuntime) {
    guard moduleInstance == nil else {
      print("[RuneSafeArea] Module already initialized")
      return
    }
    
    // Use the static initialize method from RuneSafeAreaModule
    RuneSafeAreaModule.initialize(with: runtime)
    print("[RuneSafeArea] Module initialized")
  }
  
  /// Clean up the module (called during app teardown)
  @objc public static func cleanup() {
    moduleInstance = nil
  }
}

File: packages/rune-safe-area/android/RuneSafeArea/src/main/java/dev/rune/safearea/RuneSafeArea.kt
package dev.rune.safearea

import android.app.Activity
import com.rune.kit.runtime.RuneRuntime

/**
 * Public interface for RuneSafeArea module
 */
object RuneSafeArea {
    private var moduleInstance: RuneSafeAreaModule? = null
    private var currentActivity: Activity? = null
    private var currentRuntime: RuneRuntime? = null

    /**
     * Initialize the safe area module with an Activity and RuneRuntime instance
     * Should be called during app initialization before the JS bundle loads
     */
    @JvmStatic
    fun initialize(activity: Activity, runtime: RuneRuntime) {
        if (moduleInstance != null && currentActivity === activity && currentRuntime === runtime) {
            android.util.Log.w("RuneSafeArea", "Module already initialized for this activity/runtime")
            return
        }

        if (moduleInstance != null) {
            android.util.Log.d("RuneSafeArea", "Reinitializing module for new activity/runtime")
            cleanup()
        }

        moduleInstance = RuneSafeAreaModule(activity, runtime)
        runtime.installModules(listOf(moduleInstance!!))
        currentActivity = activity
        currentRuntime = runtime
        android.util.Log.d("RuneSafeArea", "Module initialized")
    }

    /**
     * Clean up the module (called during app teardown)
     */
    @JvmStatic
    fun cleanup() {
        moduleInstance?.onDestroy()
        moduleInstance = null
        currentActivity = null
        currentRuntime = null
    }
}
